### PR TITLE
A proper implementation of String.isSwiftIdentifier

### PR DESCRIPTION
**Pull request by broadwaylamb**
Stolen from [the Swift lexer](https://github.com/apple/swift/blob/1846a5957ee971979d86bf843a0e4885da2b5a46/lib/Parse/Lexer.cpp).

### CODE REVIEW

**Comment by brentdax** [comment]
Thank you for working on this!

Before we merge this change, you will need to add tests of its functionality. What I would suggest you do is look at each identifier used in [Swift's test/Parse/identifiers.swift](https://github.com/apple/swift/blob/master/test/Parse/identifiers.swift) and determine if the test is checking that it is valid or invalid, then create a Tests/SwiftDriverTests/StringAdditionsTests.swift file with an XCTestCase that makes sure `isSwiftIdentifier` returns the appropriate result for each of them.

Once you've ported these test cases, think about if there are any others you should add. test/Parse/identifiers.swift actually doesn't seem to include some very basic tests, like identifiers containing numbers. For bonus points, you could backport these additional test cases to test/Parse/identifiers.swift, too!

To allow this test to be written, you may need to make `isSwiftIdentifier` public. Please don't use `@testable import`; we want to be able to run tests on release builds, and release builds can't enable testable import.

Thanks!

**Comment by broadwaylamb** [comment]
>To allow this test to be written, you may need to make isSwiftIdentifier public.

Do I need to add a prefix to the property, since I'm extending a type I don't own with a declaration that doesn't involve types I own?

**Comment by brentdax** [comment]
I don't think we have an official policy on prefixing extension methods, but I asked around and @aciidb0mb3r at least thinks it's a good idea. Please add an `sd_` prefix.

(Thank you for thinking of this—I sure didn't.)

**Comment by broadwaylamb** [comment]
@brentdax I've added some tests. However, I'm not sure whether we want to consider Swift keywords as identifiers here. Right now `import`, `class` etc. are "identifiers".

**Comment by brentdax** [comment]
@broadwaylamb You should include tests proving that `sd_isSwiftIdentifier` *accepts* keywords. The C++ driver uses `Lexer::isIdentifier()`, which accepts keywords, so that's the behavior we want in swift-driver too. What it's trying to do is make sure that the module name would be valid in an `import` statement, and keywords would be valid in an `import` as long as the user surrounded them with backticks, so accepting them is the right behavior here.

File: Tests/SwiftDriverTests/StringAdditionsTests.swift [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        XCTAssertTrue("var".sd_isSwiftIdentifier)
+        XCTAssertTrue("typealias".sd_isSwiftIdentifier)
+        XCTAssertTrue("class".sd_isSwiftIdentifier)
+        XCTAssertTrue("struct".sd_isSwiftIdentifier)
+        XCTAssertTrue("enum".sd_isSwiftIdentifier)
+        XCTAssertTrue("associatedtype".sd_isSwiftIdentifier)
+        XCTAssertTrue("prefix".sd_isSwiftIdentifier)
+        XCTAssertTrue("infix".sd_isSwiftIdentifier)
+        XCTAssertTrue("postfix".sd_isSwiftIdentifier)
+        XCTAssertTrue("_".sd_isSwiftIdentifier)
```
Comment (brentdax) [inline Tests/SwiftDriverTests/StringAdditionsTests.swift]: Good, you've already included these tests.

**Comment by brentdax** [comment]
@swift-ci please smoke test

**Comment by brentdax** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test

**Pull request merged**